File: 01-Languages/Go/09-Error-Handling/main.go
// main.go - (value, error) pattern, custom error types with errors.As, panic/recover.
package main

import (
	"errors"
	"fmt"
)

func validateAge(age int) (int, error) {
	if age < 0 {
		return 0, errors.New("age cannot be negative")
	}
	return age, nil
}

type ValidationError struct {
	Field   string
	Message string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("%s: %s", e.Field, e.Message)
}

func validate(age int) error {
	if age < 0 {
		return &ValidationError{Field: "age", Message: "cannot be negative"}
	}
	return nil
}

func mustDivide(a, b int) int {
	if b == 0 {
		panic("division by zero")
	}
	return a / b
}

func safeDivide(a, b int) (result int, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("recovered from panic: %v", r)
		}
	}()
	return mustDivide(a, b), nil
}

func main() {
	fmt.Println("--- (value, error) pattern ---")
	_, err := validateAge(-5)
	if err != nil {
		fmt.Println("Validation failed:", err)
	}

	fmt.Println("\n--- custom error type with errors.As ---")
	err2 := validate(-5)
	var valErr *ValidationError
	if errors.As(err2, &valErr) {
		fmt.Println("Field:", valErr.Field, "Message:", valErr.Message)
	}

	fmt.Println("\n--- panic/recover ---")
	result, err3 := safeDivide(10, 0)
	if err3 != nil {
		fmt.Println("safeDivide error:", err3)
	} else {
		fmt.Println("safeDivide result:", result)
	}

	result2, err4 := safeDivide(10, 2)
	fmt.Println("safeDivide(10, 2):", result2, "err:", err4)
}

File: 01-Languages/Rust/09-Error-Handling/main.rs
// main.rs - Result<T,E> with match, the ? operator, custom error types, panic! as last resort.

use std::fmt;

fn divide(a: f64, b: f64) -> Result<f64, String> {
    if b == 0.0 {
        return Err(String::from("cannot divide by zero"));
    }
    Ok(a / b)
}

fn calculate(a: f64, b: f64, c: f64) -> Result<f64, String> {
    let step1 = divide(a, b)?;
    let step2 = divide(step1, c)?;
    Ok(step2)
}

#[derive(Debug)]
struct ValidationError {
    field: String,
    message: String,
}

impl fmt::Display for ValidationError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}: {}", self.field, self.message)
    }
}

impl std::error::Error for ValidationError {}

fn validate_age(age: i32) -> Result<i32, ValidationError> {
    if age < 0 {
        return Err(ValidationError {
            field: String::from("age"),
            message: String::from("cannot be negative"),
        });
    }
    Ok(age)
}

fn main() {
    println!("--- Result handled via match ---");
    match divide(10.0, 0.0) {
        Ok(result) => println!("Result: {}", result),
        Err(e) => println!("Error: {}", e),
    }

    println!("\n--- ? operator propagating through a multi-step calculation ---");
    match calculate(100.0, 5.0, 0.0) {
        Ok(result) => println!("calculate succeeded: {}", result),
        Err(e) => println!("calculate failed: {}", e),
    }
    match calculate(100.0, 5.0, 2.0) {
        Ok(result) => println!("calculate succeeded: {}", result),
        Err(e) => println!("calculate failed: {}", e),
    }

    println!("\n--- custom error type implementing Display ---");
    match validate_age(-5) {
        Ok(age) => println!("Valid age: {}", age),
        Err(e) => println!("Validation failed: {}", e),
    }
}
